app/lti.go

```diff
@@ -211,9 +209,19 @@ func (a *App) generateFreeUsername(baseUsername string) (string, *model.AppError
 		return baseUsername, nil
 	}
 
+	// When using slices of strings, Go does not check that you are slicing after the last
+	// character.  Thus, we have to calculate how big of a slice we want to take.  This will
+	// be either the baseUsername length or the max username length - suffix for appended numbers
+	var sliceLen = 0
+	if (model.UserNameMaxLength - len(strconv.Itoa(MaxUsernameSuffix))) > len(baseUsername) {
+		sliceLen = len(baseUsername)
+	} else {
+		sliceLen = model.UserNameMaxLength - len(strconv.Itoa(MaxUsernameSuffix))
+	}
```

Comment:
it took me a minute to understand the condition here. i don't think this necessarily needs to be changed or anything but fwiw i find something like this clearer / easier to read
```suggestion
	// When using slices of strings, Go does not check that you are slicing after the last
	// character.  Thus, we have to calculate how big of a slice we want to take.  This will
	// be either the baseUsername length or the max username length minus the length of the largest suffix,
	// whichever is smaller
	var maxBaseUsernameLength = model.UserNameMaxLength - len(strconv.Itoa(MaxUsernameSuffix)))
	var sliceLen = math.Min(len(baseUsername), maxBaseUsernameLength)
```

Reply:
oh what the hell is go's built-in `math.Min` only for numbers of type `float64`?

Reply:
Yep, I was just seeing that.  I thought your idea and Mike's combined to a good solution!!!

Reply:
it looks like there are other parts of the mattermost that get around this by casting the values to `float64` first and then casting the result to `int64` but that's annoying

Reply:
Yeah, some other discussions say to write your own min method

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

Reply:
Will end up being
```suggestion
		var maxBaseUsernameLen = UserNameMaxLength - len(strconv.Itoa(MaxUsernameSuffix))
	var basenameLen = int64(math.Min(float64(len(baseUsername)), float64(maxBaseUsernameLen)))

	baseUsername = baseUsername[:basenameLen]
```